dd-convert.8
.\" Copyright (c) 2017 Peter J. Philipp
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 27, 2017
.Dt DD-CONVERT 8
.Os 
.Sh NAME
.Nm dd-convert
.Nd the DelphinusDNS Convert Utility
.Sh SYNOPSIS
.Nm dd-convert
.Op Fl a Ar algorithm
.Op Fl B Ar bits
.Op Fl e Ar seconds
.Op Fl h
.Op Fl i Ar inputfile
.Op Fl I Ar iterations
.Op Fl k Ar KSKfile
.Op Fl K
.Op Fl m Ar mask
.Op Fl n Ar zonename
.Op Fl o Ar outputfile
.Op Fl s Ar salt
.Op Fl t Ar ttl
.Op Fl v
.Op Fl z Ar ZSKfile
.Op Fl Z
.Sh DESCRIPTION
.Nm
is a utility which signs a delphinusdnsd zone file to output into a 
delphinusdnsd readable DNSSEC zonefile.  This utility is based on the
ruby program dd-convert.rb which was part of delphinusdns 1.0 release.  From 
delphinusdns 1.1 onwards this is the utility expected to be used for zone 
signing, unless the ruby utility proves better.
.Pp
The options are as follows:
.Pp
.Bl -tag -width Ds
.It Fl a
.Ar algorithm
specifies the algorithm to use represented as an integer.  There currently is
only 3 algorithms possible and they are of RSA types.  There is 
RSASHA1-NSEC3-SHA1
which has algorithm 7, there is RSASHA256 which has algorithm 8 (the default), 
and there is RSASHA512 which has algorithm 10.
.It Fl B
.Ar bits
Specifies the number of bits used with the algorithm.  The default value is
2048.
.It Fl e
.Ar seconds
Specifies how many seconds into the future the expiry time of the signed zone
will be.  By default this is 5184000 seconds.  To re-sign a zone with a
different signtime than what the current available time (now) is, you'll have
to recompile the binary with PROVIDED_SIGNTIME set to 1.
.It Fl h
Provides a usage output.
.It Fl i
.Ar inputfile
Specifies an inputfile, used in conjunction with the
.Fl n
option which specifies the name of the zone.
.It Fl I 
.Ar iterations
This is the number of iterations used with the NSEC3 resource record hash that 
.Nm
 produces.
.It Fl k
.Ar KSKfile
When re-signing a zone with a provided KSK keyfile use this option to specify
the KSK file.  Roll-overs are not yet supported.
.It Fl K
Sign a zone with a newly created KSK keyfile.  It is assumed that LibreSSL will
create random defaults that are cryptographically strong.
.It Fl m
.Ar mask
is a neat way to not execute components in
.Nm .
The mask values are as follows: MASK_PARSE_FILE (0x2), MASK_ADD_DNSKEY (0x4),
MASK_CONSTRUCT_NSEC3 (0x8), MASK_CALCULATE_RRSIGS (0x10), MASK_CREATE_DS (0x20)
MASK_DUMP_DB (0x40).  0x1 and 0x80 are reserved for possible BIND zone file
parsing and dumping.  As an example you may specify 42 as a mask and it would
only parse a delphinusdnsd zonefile and dump it back out as delphinusdnsd
format (without comments).  By default a mask of 0x7E exists.
.It Fl n
.Ar zonename
This option in conjunction with the 
Fl i
option must be provided.  Otherwise dd-convert will not know what to sign.
.It Fl o
.Ar outputfile
When this option is not specified dd-convert will output to standard output.  A
- provided to this option will also output to standard output otherwise it will
be output to the specified file.
.It Fl s
.Ar salt
is the salt for the NSEC3 hash.  This is a security option that applies the
same as anything that requires salt.  The input is in hexadecimal.
.It Fl t
.Ar ttl
is the time to live for created DNSKEY resource records.  Default is 3600.
.It Fl v
prints the version of 
.Nm 
and exits.
.It Fl z
.Ar ZSKfile
When re-signing a zone with a provided ZSK keyfile use this option to 
specify the KSK file.  Roll-overs are not yet supported.
.It Fl Z
Sign a zone with a newly created ZSK keyfile.  It is assumed that LibreSSL
will create random defaults that are cryptographically strong.
.El
.Sh FILES
.Pa ./tmp.XXXXXXXXXX	
a temporary database path (gets cleaned up after successful exit).
.Sh SEE ALSO
.Xr delphinusdnsd 8 , 
.Xr delphinusdns.conf 5 , 
.Sh AUTHORS
This software was written by
.An Peter J. Philipp Aq petphi@delphinusdns.org



CHANGES
Changes in RELEASE_1_3 from RELEASE_1_2
- ...

Changes in RELEASE_1_2 from RELEASE_1_1

- remove SPF support (deprecated RFC 7208 section 3.1)
- add TLSA anyreply support (affects AXFR)
- change BerkeleyDB to tree(3) RB_* macros (faster?)
- fix EDNS compliancy:  https://ednscomp.isc.org/ednscomp/5a70566c18
- fix truncation code and give empty answers with TC bit set when appropriate
- fix CNAME's with DNSSEC
- fix PTR's with DNSSEC
- move to imsg functions with privsep for IPC messaging
- OpenBSD delphinusdnsd is pledged

Changes in RELEASE_1_1 from RELEASE_1_0

- dd-convert.c written to replace dd-convert.rb.  It now supports TLSA RR's as
an added bonus.  This utility does not rely on any external (BIND) tools except
for the linkage against openssl/libressl.
- manpage for dd-convert
- Mac OS X port has been dropped
- RFC 8020 compliance with empty non-terminal names starting this release
- Linux now depends on libbsd

Changes in RELEASE_1_0 from BETA_9

- name change from wildcarddnsd to delphinusdnsd
- fixed the long standing bug that assumed one time to live for all RRSETs in a domain name (ttl fix)
- internal database has been changed in it's ABI
- wildcarding is off for good (due to DNSSEC)
- split-horizon for A records is gone, (may be added back one day)
- EDNS fixup (RFC 2671)
- DNSSEC (RFC 4033, 4034, 4035)
- DNSSEC NSEC3 (RFC 5155)
- TLSA RR (described in RFC 6698)
- dd-convert.rb ruby zone sign tool
- Linux now relies on libressl 2.2.4 library, details in the configure
- some bug fixes

Changes in BETA_9 from BETA_8

- we have replaced parse.c with parse.y
- Linux now relies on the libressl 2.1.1 library, details in the configure
- commandline options have been moved to configuration file options to coexist
- options keyword in the configfile
- SPF (RR 99) RFC 4408 support
- fix a bug related to truncation
- SSHFP (RR 44) RFC 4255 (+ RFC 6594) support
- defineable ratelimit between 1 and 127 pps , see example8.conf for sample
- EDNS0 (RFC 6891) get detected and replied upon
- NAPTR (RFC 2915) support
- Notify (RFC 1996) master-only support
- underlying BerkeleyDB databases are now stored in /var/db/wdns/[pid]/* 
  instead of /var/db/wdns, cleanup code for this was written as well.
- filter and whitelist ip ranges
- a wildcarddns.conf(5) manpage based on BNF similar to pf.conf(5) with 
  reference from parse.y
- we now log received bytes and sent bytes per query, log at the end of a
  log message.


Changes in BETA_8 from BETA_7

- we have a non-forking debug mode now with verbosity flag
- make W compile and run on the Raspberry Pi (great projects with this)

Changes in BETA_7 from BETA_6

- remote (sys)-logging feature with HMAC message verification
- change to Berkeley DB 4.6+ on BSD hosts (Linux had it already)
- allow numerous copies of wildcardnsd to pre-fork (-n flag)
- AXFR master code (this allows a bind9 slave to work with wildcarddnsd)
- fix an integer overrun in parse.c which prevented big endian hosts to
  make use of ANY replies and AXFR
- SRV RR support
- left sourceforge.net around August 14, 2012
- Mac OS X compiles now, but has not been tested with queries

Changes in BETA_6 from BETA_5

- Addition of the utterly broken recursive lookup code
- conf file is at version 4 now
- inclusion of a "root hints" entry in some example.conf's
- "ANY" RR type replies
- pidfiles written to /var/run
- master process which can be HUP'ed for reloading the nameserver
- compress_label fix from possible overflow
- DNS TTL logging fix for OpenBSD
- TCP filedescriptor fix (so -> tnp->so)


dd-convert.rb
#!/usr/bin/env ruby
#
# $Id: dd-convert.rb,v 1.9 2016/01/21 08:50:12 pjp Exp $
#
# Copyright (c) 2015 Peter J. Philipp
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. The name of the author may not be used to endorse or promote products
#    derived from this software without specific prior written permission
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
# NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
#
# nvi, ts=2
#
# This utility to delphinusdnsd takes a delphinusdnsd zonefile and converts
# it to BIND style zonefile, It then does dnssec-signzone from the BIND 
# utilities and with help of the dns-zone gem (version 0.2.0+) from Luke 
# Antins converts the signed zone back to delphinusdnsd zonefile format.  
# One day perhaps I'll write my own zonefile signer, but not today.
#

require 'fileutils'
require 'optparse'
require 'openssl'
require 'pp'
require 'tempfile'

require 'etc'
require 'dns/zone'

#
#
# Class ParseArguments, parses the commandline arguments, returns a hash
#
#

class ParseArguments < Hash
	def initialize(args)
		super()
		self[:zonename] = "centroid.eu."
		self[:numbits] = 1024 
		self[:KSK] = 0
		self[:ZSK] = 0
		self[:algorithm] = "rsasha256"
		self[:kskname] = ""
		self[:zskname] = ""
		self[:salt] = ""
		self[:ttl] = 3600
		self[:iterations] = 10
		self[:xsystem] = ""

		opts = OptionParser.new do |opts|
			opts.banner = "Usage: #$0 [options]"

			opts.on('-a', '--algorithm ALG', 'the signing algorithm, default is rsasha256') do |alg|
				if alg != "rsasha1"
					puts "illegal algorithm!"
					exit 1
				end
			end

			opts.on('-B', '--bits NUMBITS', 'the amount of bits to sign this keypair') do |bits|
				if bits > 512 && bits < 4096 then
					self[:numbits] = bits
				else
					puts "illegal bitsize for keys!"
					exit 1
				end
			end

			opts.on('-I', '--iterations num', 'iterations of NSEC3 hash') do |filename|
				self[:iterations] = filename
			end

			opts.on('-i', '--input INPUTFILE', 'the delphinusdns zonefile') do |filename|
				self[:input] = filename
				self[:bindfile] = ""
				if !File.exist?(filename) then
					puts "no such inputfile"
					exit 1
				end
			end

			opts.on('-K', '--KSK', 'create KSK keys') do |ksk|
				self[:KSK] = 1
			end

			opts.on('-k', '--kskname KEY', 'use specific KSK key') do |kskname|
				self[:kskname] = kskname || ''
			end

			opts.on('-n', '--name ZONENAME', 'the zonename of the zone to work on') do |filename|
				self[:zonename] = filename || ''
				if self[:zonename][-1] == '.' then
					puts 'trailing dot not allowed here'
					exit 1
				end
			end

			opts.on('-o', '--output OUTPUTFILE', 'the signed output zonefile') do |filename|
				self[:output] = filename
				if !File.exist?(filename) then
					puts "no such outputfile"
					exit 1
				end
			end

			opts.on('-s', '--salt SALT', 'the salt for the NSEC3') do |filename|
				self[:salt] = filename || ''
			end

			opts.on('-t', '--ttl TTL', 'the TTL for the keys') do |filename|
				self[:ttl] = filename
			end

			opts.on('-X', '--XSystem', 'specify an alternate system') do |system|
				self[:xsystem] = system || ''
			end

			opts.on('-Z', '--ZSK', 'create ZSK keys') do |zsk|
				self[:ZSK] = 1
			end

			opts.on('-z', '--zskname KEY', 'use specific ZSK key') do |zskname|
				self[:zskname] = zskname || ''
			end

			opts.on_tail('-h', '--help', 'display this help and exit') do 
				puts opts
				exit 0
			end
		end

		opts.parse!(args)
	end
end

#
#
# Class InputFile parses the input config file, returns an Array
#
#

class MyTranslation < Array
	#
	# @@name2rr is a hash that translates a given key name to a number
	#
	@@name2rr = { 	"a" 	=> 	1,
			"A" 	=> 	1,
			"ns" 	=>	2,
			"NS" 	=>	2,
			"delegate" =>	2,
			"DELEGATE" =>	2,
			"hint"	=>	2,
			"HINT"	=>	2,
			"cname" =>	5,
			"CNAME" =>	5,
			"soa"	=>	0,		# this is a hack
			"SOA"	=>	0,		# it really is 6
			"ptr"	=>	12,
			"PTR"	=>	12,
			"mx"	=>	15,
			"MX"	=>	15,
			"txt"	=>	16,
			"TXT"	=>	16,
			"aaaa"  =>	28,
			"AAAA"  =>	28,
			"srv"	=>	33,
			"SRV"	=>	33,
			"naptr" =>	35,
			"NAPTR" =>	35,
			"opt"	=>	41,
			"OPT"	=>	41,
			"ds"	=>	43,
			"DS"	=>	43,
			"sshfp"	=>	44,
			"SSHFP"	=>	44,
			"rrsig" =>	46,
			"RRSIG" =>	46,
			"nsec"	=>	47,
			"NSEC"	=>	47,
			"dnskey" =>	48,
			"DNSKEY" =>	48,
			"nsec3" =>	50,
			"NSEC3" =>	50,
			"nsec3param" => 51,
			"NSEC3PARAM" => 51,
			"spf"	=>	99,
			"SPF"	=>	99,
			"tsig"	=>	250,
			"TSIG"	=>	250,
			"ixfr"	=>	251,
			"IXFR"	=>	251,
			"axfr"	=>	252,
			"AXFR"	=>	252,
			"any"	=>	255,
			"ANY"	=>	255
		}

	#
	# @@rr2name is a hash that translates a name to a given rrtype string
	#
	@@rr2name = {
			1 	=> 	"a",
			2	=>	"ns",
			5	=>	"cname",
			6	=>	"soa",		
			12	=>	"ptr",
			15	=>	"mx",
			16	=>	"txt",
			28	=>	"aaaa",
			33	=>	"srv",
			35	=>	"naptr",
			41	=>	"opt",
			43	=>	"ds",
			44	=>	"sshfp",
			46	=>	"rrsig",
			47	=>	"nsec",
			48	=>	"dnskey",
			50	=>	"nsec3",
			51	=>	"nsec3param",
			99	=>	"spf",
			250	=>	"tsig",
			251	=>	"ixfr",
			252	=>	"axfr",
			255	=>	"any",
		}


end

class ConvertD2B < MyTranslation
	def initialize(filename)
		super()
		m = []
		ttl = 0

		linenum = 0
		filename.each() do |line|
			if line != nil then
				m = line.split(',')
				if m[1] == nil then
					next
				elsif m[1].downcase == "soa" then
					self[linenum] = '$TTL ' + m[2]
					linenum += 1
					ttl = m[2].to_i
					if m[0][-1] == "." then
						self[linenum] = '$ORIGIN ' + m[0]
						origin = m[0]
					else
						self[linenum] = '$ORIGIN ' + m[0] + '.'	
						origin = m[0] + '.'
					end
					linenum += 1
					self[linenum] = origin + " IN SOA " + m[3] 
					self[linenum] += " " + m[4] + " ( "
					self[linenum] += m[5] + " " + m[6]
					self[linenum] += " " + m[7] + " "
					self[linenum] += m[8] + " " + m[9]
					self[linenum] += " )"
					linenum += 1
				elsif m[1].downcase == "ns" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN NS " + m[3]

					linenum += 1
				elsif m[1].downcase == "a" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN A " + m[3]

					linenum += 1
				elsif m[1].downcase == "aaaa" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN AAAA " + m[3]

					linenum += 1
				elsif m[1].downcase == "cname" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN CNAME " + m[3]

					linenum += 1
				elsif m[1].downcase == "ptr" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN PTR " + m[3]

					linenum += 1
				elsif m[1].downcase == "txt" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					mytext = m[3...m.length].join(",")
					self[linenum] += " IN TXT " + mytext

					linenum += 1
				elsif m[1].downcase == "spf" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN SPF " + m[3]

					linenum += 1
				elsif m[1].downcase == "mx" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN MX " + m[3] 
					self[linenum] += " " + m[4]

					linenum += 1
				elsif m[1].downcase == "srv" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN SRV " + m[3] 
					self[linenum] += " " + m[4] + " "
					self[linenum] += m[5] + " " + m[6]

					linenum += 1
				elsif m[1].downcase == "sshfp" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN SSHFP " + m[3] 
					self[linenum] += " " + m[4] + " (" 
					m[5][0] = ' ';
					m[5][-1] = ' ';
					self[linenum] += m[5] + ")"

					linenum += 1
				elsif m[1].downcase == "naptr" then
					if m[0][-1] == "." then
						self[linenum] = m[0]
					else
						self[linenum] = m[0] + '.'	
					end
					if m[2].to_i != ttl then
						self[linenum] += " " + m[2]
					end
					self[linenum] += " IN NAPTR (" + m[3] 
					self[linenum] += " " + m[4] + " " 
				 	self[linenum] += m[5] + " " + m[6]	
					self[linenum] += " " + m[7] + " " 
				 	self[linenum] += m[8] + " )"

					linenum += 1
				end
					
			end
		end

	end
end

class InputFile < MyTranslation
	def initialize(filename, zonename)
		super()
		inputfile = []
		zonefile = []
		m = []
		savezonename = zonename
		linecount = 0

		if filename == nil
			usage
		end
		

		f = File.open(filename, "r") 
			while !f.eof do 
				inputfile << f.readline
			end
		f.close

		zone = false
		zonename = '"' + zonename + '"'
		inputfile.each() do |line| 
			m = line.split
			if m[0] == "zone" && m[1] == zonename then
				zone = true
			end
			if zone == true then
				if line =~ /^.*}/ then
					zone = false
					break
				end
				zonefile << line
			end
		end

		count = 0
		zonefile.each() do |line|
			m = self.lookup_record(zonefile, count)
			if m == nil then
				count += 1
				next
			end
			self[linecount] = m.join(",").chomp.strip
			linecount += 1
			count += 1
		end

	end

	def lookup_record(zonefile, number)
		line = zonefile[number]	

		if line =~ /zone/ || line =~ /^.*;/ || line =~ /}/ || \
			line =~ /^$/ then
			return nil
		end
		
		if line == nil then
			return nil
		end

		m = line.split(',')
		
		return m
	end
end


#
# Class MyCreateKSK - create a KSK and write it out
#
#

class MyCreateKeys < Hash
	#
	# taken from:
	# http://www.iana.org/assignments/dns-sec-alg-numbers/dns-sec-alg-numbers.xhtml
	# may be updated
	# we only support 2 of these rsasha1 and rsasha256
	#
	@@alg2num = {
			"rsamd5"		=>	1,
			"dh"			=>	2,
			"dsa"			=>	3,
			"rsasha1"		=>	5,
			"dsa-nsec3-sha1"	=>	6,
			"rsasha1-nsec3-sha1"	=>	7,
			"rsasha256"		=>	8,
			"rsasha512"		=>	10,
			"ecc-gost"		=>	12,
			"ecdsap256sha256"	=>	13,
			"ecdsap384sha384"	=>	14
		}

	def initialize(type, algorithm, bits, zonename, ttl, xsystem)
		self[:zskname] = ''
		self[:kskname] = ''
		systemid = [] 
		super()

		if xsystem == "" then
			systemid = Etc.uname[:sysname]
		else
			systemid = xsystem
		end
		
		if type == 1 then
			createKSK(algorithm, bits, zonename, ttl, systemid)	
		else
			createZSK(algorithm, bits, zonename, ttl, systemid)
		end
	end

	def createKSK(algorithm, bits, zonename, ttl, systemid)
		if systemid == "OpenBSD" then
			keygen = "/usr/local/sbin/dnssec-keygen"
		else
			keygen = "/usr/sbin/dnssec-keygen"
		end
		
		IO.popen(keygen + ' -3 -L ' + \
			ttl.to_s + ' -f KSK -a ' + algorithm + \
			' -b ' + bits.to_s + ' -n zone ' + \
			zonename, 'r+') do |pipe|

			self[:kskname] = pipe.read
		end

	end

	def createZSK(algorithm, bits, zonename, ttl, systemid)
		if systemid == "OpenBSD" then
			keygen = "/usr/local/sbin/dnssec-keygen"
		else
			keygen = "/usr/sbin/dnssec-keygen"
		end
		
		IO.popen(keygen + ' -3 -L ' + \
			ttl.to_s + ' -a ' + algorithm + \
			' -b ' + bits.to_s + ' -n zone ' + \
			zonename, 'r+') do |pipe|

			self[:zskname] = pipe.read
		end
	end
end

#
# usage
#
def usage
	ParseArguments.new(Hash.new("--help"));
	exit 1
end


#
# start 
#

systemid = [] 
rrline = []
arguments = ParseArguments.new(ARGV)

if arguments[:xsystem] == "" then
	systemid = Etc.uname[:sysname]
else
	systemid = arguments[:xsystem]
end

if arguments[:input] != "" then
	puts '################################################################'
	inputfile = InputFile.new(arguments[:input], arguments[:zonename])
	#puts inputfile
	bindfile = ConvertD2B.new(inputfile)
	#puts bindfile
	out = Tempfile.new('master.' + arguments[:zonename]);
	bindfile.to_a.each() do |line|
		out.puts line
	end

	if arguments[:KSK] == 1 then
		arguments[:kskname] = MyCreateKeys.new(1 , \
		arguments[:algorithm], arguments[:numbits], \
		arguments[:zonename], arguments[:ttl], \
		arguments[:xsystem])[:kskname]
			puts 'created KSK key ' + arguments[:kskname]
	end

	if arguments[:ZSK] == 1 then
		arguments[:zskname] = MyCreateKeys.new(2 , \
		arguments[:algorithm], arguments[:numbits], \
		arguments[:zonename], arguments[:ttl], \
		arguments[:xsystem])[:zskname]
			puts 'created ZSK key ' + arguments[:zskname]
	end

	if arguments[:kskname] == '' then
		puts 'unknown KSK'
		exit 1
	end

	if arguments[:zskname] == '' then
		puts 'unknown ZSK'
		exit 1
	end
	
	open(arguments[:kskname].chomp + '.key') { |f| out.puts f.read }
	open(arguments[:zskname].chomp + '.key') { |f| out.puts f.read }

	out.rewind

	if systemid == "OpenBSD" then
		signzonepath = "/usr/local/sbin/dnssec-signzone"
	elsif systemid == "FreeBSD"
		signzonepath = "/usr/sbin/dnssec-signzone"
	else
		signzonepath = "/usr/sbin/dnssec-signzone"
	end

	
	IO.popen(signzonepath + ' -O full -3 \'' + \
		arguments[:salt] + '\' -H ' + arguments[:iterations].to_s + \
		' -e +5184000 -o ' + arguments[:zonename] + '. -t  -k ' + \
		arguments[:kskname].chomp + ' ' + out.path + ' ' + \
		arguments[:zskname].chomp , 'r+') do |pipe|

		puts pipe.read
	end

	signed = out.path + '.signed'
	out.close
		
	zone = DNS::Zone.load(open(signed))
	File.delete signed

	if arguments[:output] != '' then
		output = open(arguments[:zonename] + '.signed', 'w+');
		output.puts '; this file is automatically generated, do NOT edit'
		output.puts '; it was generated by dd-convert.rb'
		output.puts 'zone "' + arguments[:zonename] + '" {'
	else
		puts 'You didn\'t specify an output file...exit.'
		exit 0
	end

	zone.records.each() do |record|
		field = record.to_s.split(":");
		
		type = field[6]
		if type == "SOA" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'soa' << dnsfield[1]
			rrline << record.nameserver << record.email.to_s
			rrline << record.serial.to_s << record.refresh_ttl.to_s 
			rrline << record.retry_ttl.to_s 
			rrline << record.expiry_ttl.to_s 
			rrline << record.minimum_ttl.to_s
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "A" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'a' << dnsfield[1]
			rrline << record.address
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "AAAA" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'aaaa' << dnsfield[1]
			rrline << record.address
			output.puts '  ' + rrline.join(",")
			rrline = [] 
		elsif type == "TXT" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'txt' << dnsfield[1]
			rrline << '"' + record.text + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "SPF" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'spf' << dnsfield[1]
			rrline << '"' + record.text + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "RRSIG" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'rrsig' << dnsfield[1]
			rrline << record.type_covered << record.algorithm.to_s 
			rrline << record.labels.to_s << record.original_ttl.to_s
			rrline << record.signature_expiration.to_s 
			rrline << record.signature_inception.to_s 
			rrline << record.key_tag.to_s 
			rrline << record.signer.to_s.gsub(/ /, "") 
			rrline << '"' + record.signature.to_s.gsub(/ /, "") \
				+ '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "NSEC3PARAM" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'nsec3param' << dnsfield[1]
			rrline << record.algorithm.to_s 
			rrline << record.flags.to_s 
			rrline << record.iterations.to_s
			rrline << '"' + record.salt.to_s + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "NSEC3" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'nsec3' << dnsfield[1]
			rrline << record.algorithm.to_s 
			rrline << record.flags.to_s 
			rrline << record.iterations.to_s
			rrline << '"' + record.salt.to_s + '"'
			rrline << '"' + record.next_hashed_owner_name.to_s \
				+ '"'
			rrline << '"' + record.rrset_types.to_s + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "DNSKEY" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'dnskey' << dnsfield[1]
			rrline << record.flags.to_s
			rrline << record.protocol.to_s
			rrline << record.algorithm.to_s
			rrline << '"' + record.key.to_s.gsub(/ /, "") + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "DS" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'ds' << dnsfield[1]
			rrline << record.key_tag.to_s
			rrline << record.algorithm.to_s
			rrline << record.digest_type.to_s
			rrline << '"' + record.digest.to_s.gsub(/ /, "") + '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "MX" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'mx' << dnsfield[1]
			rrline << record.priority.to_s << record.exchange.to_s
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "NS" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'ns' << dnsfield[1]
			rrline << record.nameserver
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "CNAME" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'cname' << dnsfield[1]
			rrline << record.domainname	
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "PTR" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'ptr' << dnsfield[1]
			rrline << record.name
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "NAPTR" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'naptr' << dnsfield[1]
			rrline << record.order.to_s 
			rrline << record.pref.to_s
			rrline << '"' + record.service + '"'
			rrline << '"' + record.regexp + '"'
			rrline << record.replacement.to_s
			# this is mislabelled in dns-zone! but preserve order
			rrline << record.flags
			#
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "SRV" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'srv' << dnsfield[1]
			rrline << record.priority.to_s
			rrline << record.weight.to_s << record.port.to_s
			rrline << record.target.to_s 	
			output.puts '  ' + rrline.join(",")
			rrline = []
		elsif type == "SSHFP" then
			dnsfield = record.general_prefix
			rrline << dnsfield[0] << 'sshfp' << dnsfield[1]
			rrline << record.algorithm_number.to_s
			rrline << record.fingerprint_type.to_s
			rrline << '"' + record.fingerprint.to_s.gsub(/ /, "") \
				+ '"'
			output.puts '  ' + rrline.join(",")
			rrline = []
		end
	end

	output.puts '}'

	puts 'DS set is called dsset-' + arguments[:zonename] + '.'
	puts 'Signed Zone is called ' + arguments[:zonename] + '.signed, enjoy!'
	puts '################################################################'

	exit 0
end

if arguments[:bindfile] != "" then
	inputfile = BindInputFile.new(arguments[:bindfile], arguments[:zonename])
	puts inputfile
end

exit 0
